File: hphp/hack/src/hhbc/ast_class_expr.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel
open Hhbc_ast
module A = Aast
module SN = Naming_special_names
module SU = Hhbc_string_utils

(* Class expressions; similar to Nast.class_id *)
type class_expr =
  | Class_special of SpecialClsRef.t
  | Class_id of Ast_defs.id
  | Class_expr of Tast.expr
  | Class_reified of Instruction_sequence.t

let get_original_class_name ~resolve_self ~check_traits scope =
  match Ast_scope.Scope.get_class scope with
  | None -> None
  | Some cd ->
    if (cd.A.c_kind = Ast_defs.Ctrait && not check_traits) || not resolve_self
    then
      None
    else
      let class_name = snd cd.A.c_name in
      (* If we're in a closure we unmangle the name to get the original class *)
      (match SU.Closures.unmangle_closure class_name with
      | None -> Some class_name
      | Some _ when resolve_self ->
        begin
          match
            SMap.find_opt class_name (Emit_env.get_closure_enclosing_classes ())
          with
          | Some cd when cd.Emit_env.kind <> Ast_defs.Ctrait ->
            Some cd.Emit_env.name
          | _ -> None
        end
      | _ -> None)

let get_parent_class_name cd =
  match cd.A.c_extends with
  | [(_, A.Happly ((_, parent_cid), _))] -> Some parent_cid
  | _ -> None

let get_original_parent_class_name ~check_traits ~resolve_self scope =
  match Ast_scope.Scope.get_class scope with
  | None -> None
  | Some cd ->
    (* Parent doesn't make sense on an interface, so we treat
      "parent" as the literal class id to grab from *)
    if cd.A.c_kind = Ast_defs.Cinterface then
      Some SN.Classes.cParent
    else if
      (cd.A.c_kind = Ast_defs.Ctrait && not check_traits) || not resolve_self
    then
      None
    else
      let class_name = snd cd.A.c_name in
      (match SU.Closures.unmangle_closure class_name with
      | Some _ ->
        let cd_opt =
          SMap.find_opt class_name (Emit_env.get_closure_enclosing_classes ())
        in
        Option.bind cd_opt (fun x -> x.Emit_env.parent_class_name)
      | None -> get_parent_class_name cd)

let expr_to_class_expr
    ?(check_traits = false) ~resolve_self scope ((_, expr_) as expr) =
  match expr_ with
  | A.Id (_, id) when SU.is_static id -> Class_special SpecialClsRef.Static
  | A.Id (pos, id) when SU.is_parent id ->
    begin
      match
        get_original_parent_class_name ~resolve_self ~check_traits scope
      with
      | Some name -> Class_id (pos, name)
      | None -> Class_special SpecialClsRef.Parent
    end
  | A.Id (pos, id) when SU.is_self id ->
    begin
      match get_original_class_name ~resolve_self ~check_traits scope with
      | Some name -> Class_id (pos, name)
      | None -> Class_special SpecialClsRef.Self
    end
  | A.Id id -> Class_id id
  | _ -> Class_expr expr

(* The legacy AST used expr node everywhere where the AAST uses a class_id node.
 * The ast_to_nast module translates all of these occurrences to CIexpr nodes
 * so always expect CIexpr as the class_id node
 *)
let class_id_to_class_expr
    ?(check_traits = false) ~resolve_self scope (cid : Tast.class_id) :
    class_expr =
  let (annot, cid_) = cid in
  let e =
    match cid_ with
    | A.CIexpr e -> e
    | A.CI sid -> (annot, A.Id sid)
    | _ -> failwith "TODO Unimplemented class_id_to_class_expr"
  in
  expr_to_class_expr ~check_traits ~resolve_self scope e

File: hphp/hack/src/hhbc/ast_class_expr.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use ast_scope_rust::{self as ast_scope, Scope};
use env::emitter::Emitter;
use hhbc_ast_rust::SpecialClsRef;
use hhbc_string_utils_rust as string_utils;
use instruction_sequence::InstrSeq;
use naming_special_names_rust::classes;
use oxidized::{aast::*, ast, ast_defs};

#[derive(Debug)]
pub enum ClassExpr {
    Special(SpecialClsRef),
    Id(ast_defs::Id),
    Expr(ast::Expr),
    Reified(InstrSeq),
}

impl ClassExpr {
    fn get_original_class_name(
        emitter: &Emitter,
        check_traits: bool,
        resolve_self: bool,
        opt_class_info: Option<(ast_defs::ClassKind, &str)>,
    ) -> Option<String> {
        if let Some((kind, class_name)) = opt_class_info {
            if (kind != ast_defs::ClassKind::Ctrait || check_traits) && resolve_self {
                if string_utils::closures::unmangle_closure(class_name).is_none() {
                    return Some(class_name.to_string());
                } else if let Some(c) = emitter
                    .emit_global_state()
                    .get_closure_enclosing_class(class_name)
                {
                    if c.kind != ast_defs::ClassKind::Ctrait {
                        return Some(c.name.clone());
                    }
                }
            }
        }
        None
    }

    pub fn get_parent_class_name(class: &ast_scope::Class) -> Option<String> {
        if let [Hint(_, hint)] = class.get_extends() {
            if let Hint_::Happly(ast_defs::Id(_, parent_cid), _) = &**hint {
                return Some(parent_cid.to_string());
            }
        }
        None
    }

    fn get_original_parent_class_name(
        emitter: &Emitter,
        check_traits: bool,
        resolve_self: bool,
        opt_class_info: Option<(ast_defs::ClassKind, &str)>,
        opt_parent_name: Option<String>,
    ) -> Option<String> {
        if let Some((kind, class_name)) = opt_class_info {
            if kind == ast_defs::ClassKind::Cinterface {
                return Some(classes::PARENT.to_string());
            };
            if (kind != ast_defs::ClassKind::Ctrait || check_traits) && resolve_self {
                if string_utils::closures::unmangle_closure(class_name).is_none() {
                    return opt_parent_name;
                } else if let Some(c) = emitter
                    .emit_global_state()
                    .get_closure_enclosing_class(class_name)
                {
                    return c.parent_class_name.clone();
                }
            }
        }
        None
    }

    pub fn expr_to_class_expr(
        emitter: &Emitter,
        check_traits: bool,
        resolve_self: bool,
        scope: &Scope,
        expr: ast::Expr,
    ) -> Self {
        if let Some(cd) = scope.get_class() {
            Self::expr_to_class_expr_(
                emitter,
                check_traits,
                resolve_self,
                Some((cd.get_kind(), cd.get_name_str())),
                Self::get_parent_class_name(cd),
                expr,
            )
        } else {
            Self::expr_to_class_expr_(emitter, check_traits, resolve_self, None, None, expr)
        }
    }

    pub fn expr_to_class_expr_(
        emitter: &Emitter,
        check_traits: bool,
        resolve_self: bool,
        opt_class_info: Option<(ast_defs::ClassKind, &str)>,
        opt_parent_name: Option<String>,
        expr: ast::Expr,
    ) -> Self {
        match expr.1 {
            Expr_::Id(x) => {
                let ast_defs::Id(pos, id) = *x;
                if string_utils::is_static(&id) {
                    Self::Special(SpecialClsRef::Static)
                } else if string_utils::is_parent(&id) {
                    match Self::get_original_parent_class_name(
                        emitter,
                        check_traits,
                        resolve_self,
                        opt_class_info,
                        opt_parent_name,
                    ) {
                        Some(name) => Self::Id(ast_defs::Id(pos, name)),
                        None => Self::Special(SpecialClsRef::Parent),
                    }
                } else if string_utils::is_self(&id) {
                    match Self::get_original_class_name(
                        emitter,
                        check_traits,
                        resolve_self,
                        opt_class_info,
                    ) {
                        Some(name) => Self::Id(ast_defs::Id(pos, name)),
                        None => Self::Special(SpecialClsRef::Self_),
                    }
                } else {
                    Self::Id(ast_defs::Id(pos, id))
                }
            }
            _ => Self::Expr(expr),
        }
    }

    pub fn class_id_to_class_expr(
        emitter: &Emitter,
        check_traits: bool,
        resolve_self: bool,
        scope: &Scope,
        cid: &ast::ClassId,
    ) -> Self {
        let ClassId(annot, cid_) = cid;
        let expr = match cid_ {
            ClassId_::CIexpr(e) => e.clone(),
            ClassId_::CI(sid) => Expr(annot.clone(), Expr_::mk_id(sid.clone())),
            ClassId_::CIparent => return Self::Special(SpecialClsRef::Parent),
            ClassId_::CIstatic => return Self::Special(SpecialClsRef::Static),
            ClassId_::CIself => return Self::Special(SpecialClsRef::Self_),
        };
        Self::expr_to_class_expr(emitter, check_traits, resolve_self, scope, expr)
    }
}
